apps/vaporgui/VaporWidgets.cpp

```diff
@@ -0,0 +1,206 @@
+#include <QFileDialog>
+
+#include "VaporWidgets.h"
+#include "FileOperationChecker.h"
+#include "ErrorReporter.h"
+
+#include "ui_VPushButton.h"
+#include "ui_VCheckBox.h"
+#include "ui_VComboBox.h"
+#include "ui_VPathSelector.h"
+
+#include <iostream>
+#include <cassert>
+
+VPushButton::VPushButton(
+        QWidget *parent,
+        std::string labelText,
+        std::string buttonText
+    ) :
+    QWidget(parent),
+    Ui_VPushButton()
+{
+    setupUi(this);
+
+    SetLabelText( QString::fromStdString( labelText ) );
+    SetButtonText( QString::fromStdString( buttonText ) );
+
+    connect( _myButton, SIGNAL( pressed() ),
+        this, SLOT( _buttonPressed() ) );
+}
+
+void VPushButton::SetLabelText( const std::string text )
+{
+    SetLabelText(QString::fromStdString( text ) );
+}
+
+void VPushButton::SetLabelText( const QString text )
+{
+    _myLabel->setText( text );
+}
+
+void VPushButton::SetButtonText( const std::string text )
+{
+    SetButtonText(QString::fromStdString( text ) );
+}
+
+void VPushButton::SetButtonText( const QString text )
+{
+    _myButton->setText( text );
+}
+
+void VPushButton::_buttonPressed() {
+    emit _pressed();
+}
+
+VComboBox::VComboBox(
+        QWidget *parent,
+        std::string labelText
+    ) :
+    QWidget(parent),
+    Ui_VComboBox()
+{
+    setupUi(this);
+
+    SetLabelText( labelText );
+
+    connect( _myCombo, SIGNAL( currentIndexChanged(int) ),
+        this, SLOT( _userIndexChanged(int) ) );
+}
+
+void VComboBox::SetLabelText( std::string text ) {
+    _myLabel->setText( QString::fromStdString(text) );
+}
+
+void VComboBox::_userIndexChanged(int index) {
+    emit _indexChanged(index);
+}
+
+int VComboBox::GetCurrentIndex() const {
+    return _myCombo->currentIndex();
+}
+
+std::string VComboBox::GetCurrentText() const {
+    return _myCombo->currentText().toStdString();
+}
+
+void VComboBox::AddOption( std::string option, int index) {
+    _myCombo->insertItem( index, QString::fromStdString(option) );
+}
+
+void VComboBox::RemoveOption( int index=0 ) {
+    _myCombo->removeItem( index ) ;
```

Comment:
what if `index` passed in is out of bound?

Reply:
If index is out of bounds, nothing happens.

Reply:
Should this function return an error code to notify the caller if "nothing happens"?
(Brainstorming)

Reply:
I don't know.  The Qt guys didn't seem to think it's needed.  We can if you like though.

Reply:
@clyne  what do you think on this topic? I.e., should a composite widget of our own check if the indices passed in are valid?

Reply:
I'm not sure any purpose is served by informing the user of the class that they have tried to delete something that doesn't exist. I.e. I think silently ignoring an invalid request is reasonable in this case.

Reply:
Let's not check out of bound then.